From fe48b57945cc6d7dbe5014be8cd264da6f53fb37 Mon Sep 17 00:00:00 2001
From: Alvaro Guzman <alvaro.guzman@owasys.com>
Date: Mon, 24 Apr 2023 12:06:20 +0200
Subject: [PATCH] Fixing reboot issue upgrade available init value is 0

---
 include/configs/imx8mp_owa5x.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/configs/imx8mp_owa5x.h b/include/configs/imx8mp_owa5x.h
index d10bf19259..e49eca6047 100644
--- a/include/configs/imx8mp_owa5x.h
+++ b/include/configs/imx8mp_owa5x.h
@@ -189,7 +189,7 @@
 	"fdt_addr=0x88000000\0"			\
 	"fdt_high=0xffffffffffffffff\0" \
 	"fdt_file=" CONFIG_DEFAULT_FDT_FILE "\0" \
-	"upgrade_available=1\0" \
+	"upgrade_available=0\0" \
 	"boot_fit=no\0" \
 	"mtdids="	CONFIG_MTDIDS_DEFAULT	"\0" \
 	"mtdparts=" CONFIG_MTDPARTS_DEFAULT "\0" \
-- 
2.25.1

