From ed7dab1759e0c64a6543598da2da71bf51291b9d Mon Sep 17 00:00:00 2001
From: Alvaro Guzman <alvaro.guzman@owasys.com>
Date: Tue, 21 Mar 2023 17:21:52 +0100
Subject: [PATCH] Fixed set balena env variables

---
 include/configs/imx8mp_owa5x.h | 20 ++++++++++++++++----
 1 file changed, 16 insertions(+), 4 deletions(-)

diff --git a/include/configs/imx8mp_owa5x.h b/include/configs/imx8mp_owa5x.h
index 4ca024ed22..ba6b885e24 100644
--- a/include/configs/imx8mp_owa5x.h
+++ b/include/configs/imx8mp_owa5x.h
@@ -193,15 +193,27 @@
 	"boot_fit=no\0" \
 	"mtdids="	CONFIG_MTDIDS_DEFAULT	"\0" \
 	"mtdparts=" CONFIG_MTDPARTS_DEFAULT "\0" \
-	"console=ttymxc0,115200 earlycon=ec_imx6q,0x30860000,115200\0" \
+	"console=ttymxc0,115200\0" \
 	"mmcdev=2\0" \
 	"mmcpart=1\0" \
 	"mmcroot=" CONFIG_MMCROOT " rootwait quiet rw\0" \
 	"mmcautodetect=yes\0" \
 	"zip_addr=0x70480000\0" \
-	"mmcargs=setenv bootargs console=${console} root=${mmcroot} rootfstype=ext4\0" \
-	"BOOT_CMD_BALENA=run mmcargs; fatload mmc ${mmcdev}:1 ${zip_addr} ${image}; fatload mmc ${mmcdev}:1 ${fdt_addr} ${fdt_file}; unzip ${zip_addr} ${loadaddr}; "\
-	"booti ${loadaddr} - ${fdt_addr};\0" \
+	"mmcargs=setenv bootargs ${jh_clk} console=${console} ${resin_kernel_root} ${os_cmdline}\0" \
+	"set_up_balena_env_vars=setenv resin_kernel_load_addr ${loadaddr};" \
+		"run resin_set_kernel_root; run set_os_cmdline; " \
+		"setenv usbdev ${resin_dev_index};" \
+		"setenv usbbootpart ${resin_boot_part};" \
+		"setenv mmcdev ${resin_dev_index};" \
+		"setenv mmcbootpart ${resin_boot_part};\0" \
+	"balena_load_kernel=fatload mmc ${mmcdev}:${mmcbootpart} ${zip_addr} ${image}\0 " \
+	"balena_load_fdt=fatload mmc ${mmcdev}:${mmcbootpart} ${fdt_addr} ${fdt_file}\0 " \
+	"BOOT_CMD_BALENA=run set_up_balena_env_vars; " \
+		"run mmcargs; " \ 
+		"run balena_load_kernel; " \
+		"run balena_load_fdt; " \
+		"unzip ${zip_addr} ${loadaddr}; " \
+		"booti ${loadaddr} - ${fdt_addr};\0" \
 	"startmmcboot=echo Booting from mmc ...; " \
 		"run mmcargs; " \
 		"if test ${boot_fit} = yes || test ${boot_fit} = try; then " \
-- 
2.25.1

