From fc1662f72233c0a5e620cd905adf9656ea6008b2 Mon Sep 17 00:00:00 2001
From: Alvaro Guzman <alvaro.guzman@owasys.com>
Date: Wed, 14 Sep 2022 11:20:01 +0200
Subject: [PATCH 4/4] Changed Balena boot command to boot from internal memoy

---
 include/configs/imx8mp_owa5x.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/configs/imx8mp_owa5x.h b/include/configs/imx8mp_owa5x.h
index 9dcbb762..cdf35d8e 100644
--- a/include/configs/imx8mp_owa5x.h
+++ b/include/configs/imx8mp_owa5x.h
@@ -200,7 +200,7 @@
 	"mmcautodetect=yes\0" \
 	"zip_addr=0x70480000\0" \
 	"mmcargs=setenv bootargs console=${console} root=${mmcroot} rootfstype=ext4\0" \
-	"BOOT_CMD_BALENA=run mmcargs; fatload mmc 1:1 ${zip_addr} ${image}; fatload mmc 1:1 ${fdt_addr} ${fdt_file}; unzip ${zip_addr} ${loadaddr}; "\
+	"BOOT_CMD_BALENA=run mmcargs; fatload mmc 2:1 ${zip_addr} ${image}; fatload mmc 2:1 ${fdt_addr} ${fdt_file}; unzip ${zip_addr} ${loadaddr}; "\
 	"booti ${loadaddr} - ${fdt_addr};\0" \
 	"startmmcboot=echo Booting from mmc ...; " \
 		"run mmcargs; " \
-- 
2.25.1

