From f9d1feafe31a95c08fd00678461a429d2f5e265b Mon Sep 17 00:00:00 2001
From: Alvaro Guzman <alvaro.guzman@owasys.com>
Date: Tue, 9 Aug 2022 17:14:43 +0200
Subject: [PATCH 04/11] First unzip the kernel

---
 include/configs/imx8mp_owa5x.h | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/include/configs/imx8mp_owa5x.h b/include/configs/imx8mp_owa5x.h
index ae560076..1abc4849 100644
--- a/include/configs/imx8mp_owa5x.h
+++ b/include/configs/imx8mp_owa5x.h
@@ -198,8 +198,10 @@
 	"mmcpart=1\0" \
 	"mmcroot=" CONFIG_MMCROOT " rootwait quiet rw mtdparts\0" \
 	"mmcautodetect=yes\0" \
+	"zip_addr=0x70480000\0" \
 	"mmcargs=setenv bootargs ${jh_clk} console=${console} root=${mmcroot}\0 " \
-	"BOOT_CMD_BALENA=run mmcargs; fatload mmc 1:1 ${loadaddr} ${image}; fatload mmc 1:1 ${fdt_addr} ${fdt_file}; bootz ${loadaddr} - ${fdt_addr};\0" \
+	"BOOT_CMD_BALENA=run mmcargs; fatload mmc 1:1 ${zip_addr} ${image}; fatload mmc 1:1 ${fdt_addr} ${fdt_file}; unzip ${zip_addr} ${loadaddr}; "\
+	"booti ${zip_addr} - ${fdt_addr};\0" \
 	"startmmcboot=echo Booting from mmc ...; " \
 		"run mmcargs; " \
 		"if test ${boot_fit} = yes || test ${boot_fit} = try; then " \
-- 
2.25.1

