From 026bd0d94da4d658be5d465afa6a7cd9361eeaff Mon Sep 17 00:00:00 2001
From: Alvaro Guzman <alvaro.guzman@owasys.com>
Date: Wed, 10 Aug 2022 10:21:33 +0200
Subject: [PATCH 05/11] Fixed booti bug command

---
 include/configs/imx8mp_owa5x.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/configs/imx8mp_owa5x.h b/include/configs/imx8mp_owa5x.h
index 1abc4849..5208d749 100644
--- a/include/configs/imx8mp_owa5x.h
+++ b/include/configs/imx8mp_owa5x.h
@@ -201,7 +201,7 @@
 	"zip_addr=0x70480000\0" \
 	"mmcargs=setenv bootargs ${jh_clk} console=${console} root=${mmcroot}\0 " \
 	"BOOT_CMD_BALENA=run mmcargs; fatload mmc 1:1 ${zip_addr} ${image}; fatload mmc 1:1 ${fdt_addr} ${fdt_file}; unzip ${zip_addr} ${loadaddr}; "\
-	"booti ${zip_addr} - ${fdt_addr};\0" \
+	"booti ${loadaddr} - ${fdt_addr};\0" \
 	"startmmcboot=echo Booting from mmc ...; " \
 		"run mmcargs; " \
 		"if test ${boot_fit} = yes || test ${boot_fit} = try; then " \
-- 
2.25.1

