From e3de6981923c9e1455ced249a91f2da4eb74daf2 Mon Sep 17 00:00:00 2001
From: Alvaro Guzman <alvaro.guzman@owasys.com>
Date: Wed, 10 Aug 2022 10:41:31 +0200
Subject: [PATCH 06/11] bootargs fixed

---
 include/configs/imx8mp_owa5x.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/configs/imx8mp_owa5x.h b/include/configs/imx8mp_owa5x.h
index 5208d749..293a14db 100644
--- a/include/configs/imx8mp_owa5x.h
+++ b/include/configs/imx8mp_owa5x.h
@@ -199,7 +199,7 @@
 	"mmcroot=" CONFIG_MMCROOT " rootwait quiet rw mtdparts\0" \
 	"mmcautodetect=yes\0" \
 	"zip_addr=0x70480000\0" \
-	"mmcargs=setenv bootargs ${jh_clk} console=${console} root=${mmcroot}\0 " \
+	"mmcargs=setenv bootargs console=${console} root=${mmcroot}\0 "
 	"BOOT_CMD_BALENA=run mmcargs; fatload mmc 1:1 ${zip_addr} ${image}; fatload mmc 1:1 ${fdt_addr} ${fdt_file}; unzip ${zip_addr} ${loadaddr}; "\
 	"booti ${loadaddr} - ${fdt_addr};\0" \
 	"startmmcboot=echo Booting from mmc ...; " \
-- 
2.25.1

