From 93e19d629e377cb938e1fb9397f900e4fa1f3e4a Mon Sep 17 00:00:00 2001
From: Alvaro Guzman <alvaro.guzman@owasys.com>
Date: Mon, 5 Dec 2022 13:15:29 +0100
Subject: [PATCH 7/7] Kernel loaded from resin_rootA/boot folder

---
 include/configs/imx8mp_owa5x.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/configs/imx8mp_owa5x.h b/include/configs/imx8mp_owa5x.h
index 4ca024ed22..19cdb34374 100644
--- a/include/configs/imx8mp_owa5x.h
+++ b/include/configs/imx8mp_owa5x.h
@@ -200,7 +200,7 @@
 	"mmcautodetect=yes\0" \
 	"zip_addr=0x70480000\0" \
 	"mmcargs=setenv bootargs console=${console} root=${mmcroot} rootfstype=ext4\0" \
-	"BOOT_CMD_BALENA=run mmcargs; fatload mmc ${mmcdev}:1 ${zip_addr} ${image}; fatload mmc ${mmcdev}:1 ${fdt_addr} ${fdt_file}; unzip ${zip_addr} ${loadaddr}; "\
+	"BOOT_CMD_BALENA=run mmcargs; ext4load mmc ${mmcdev}:2 ${zip_addr} boot/${image}; fatload mmc ${mmcdev}:1 ${fdt_addr} ${fdt_file}; unzip ${zip_addr} ${loadaddr}; booti ${loadaddr} - ${fdt_addr}; "\
 	"booti ${loadaddr} - ${fdt_addr};\0" \
 	"startmmcboot=echo Booting from mmc ...; " \
 		"run mmcargs; " \
-- 
2.25.1

